test/any/datastore/updater-integration.es6.js

```diff
@@ -0,0 +1,182 @@
+      of: VersionedItem,
+      delegate: MDAO.create({of: VersionedItem}),
+      syncRecordDAO: MDAO.create({of: VersionedSyncRecord}),
+      remoteDAO: VersionNoDAO.create({
+        delegate: datastoreDAO,
+      }),
+      polling: true,
+    });
+    updater = DatastoreUpdater.create();
+  });
+
+  it('should only perform necessary put/remove updates', done => {
+    srcDAO.put(Item.create({id: 1, data: 'src1'})); // Newly added.
+    srcDAO.put(Item.create({id: 2, data: 'src2'})); // Updated.
+    srcDAO.put(Item.create({id: 3, data: 'same3'})); // Not updated.
+    // Bring back previously deleted record.
+    srcDAO.put(Item.create({id: 4, data: 'same4deletedToUndeleted'}));
+    // Bring back previously deleted record, and update data.
+    srcDAO.put(Item.create({id: 5, data: 'different5deletedToUndeleted'}));
+
+    // Promise returned from SyncDAO does not wait for underlying VersionNoDAO
+    // to complete its operation. Use counters to support waiting for data to
+    // arrive.
+    let dstOpsStarted = 0;
+    let dstOpsFinished = 0;
+    function dstPut(o) {
+      dstDAO.put(o);
+      dstOpsStarted++;
+    }
+    function dstRemove(o) {
+      dstDAO.put(o);
+      dstOpsStarted++;
+    }
+    datastoreDAO.listen(QuickSink.create({
+      putFn: function() { dstOpsFinished++; },
+      removeFn: function() { dstOpsFinished++; },
+    }));
+    function dstWait() {
+      return new Promise((resolve, reject) => {
+        const intervalId = setInterval(() => {
+          if (dstOpsStarted === dstOpsFinished) {
+            clearInterval(intervalId);
+            resolve();
+          }
+        }, 1);
+      });
+    }
+
+    dstPut(VersionedItem.create({
+      id: 2,
+      data: 'dst2',
+    }));
+    dstPut(VersionedItem.create({
+      id: 3,
+      data: 'same3',
+    }));
+    dstPut(VersionedItem.create({
+      id: 4,
+      data: 'same4deletedToUndeleted',
+    }));
+    dstPut(VersionedItem.create({
+      id: 5,
+      data: 'unlike5deletedToUndeleted',
+    }));
+    dstPut(VersionedItem.create({
+      id: 6,
+      data: 'toBeDeleted6',
+    }));
+    dstPut(VersionedItem.create({
+      id: 7,
+      data: 'toBeStillDeleted7',
+    }));
+    let same3version;
+    let deleted7version;
+    dstDAO.sync().then(() => dstWait())
+      .then(() => {
+        dstRemove(VersionedItem.create({
+          id: 4,
+          data: 'same4deletedToUndeleted',
+        }));
+        dstRemove(VersionedItem.create({
+          id: 5,
+          data: 'unlike5deletedToUndeleted',
+        }));
+        dstPut(VersionedItem.create({
+          id: 7,
+          data: 'toBeStillDeleted7',
+        }));
+        return dstDAO.sync().then(() => dstWait());
+      }).then(() => datastoreDAO.find(3))
+      .then(same3 => same3version = same3.version_)
+      .then(() => datastoreDAO.find(7))
+      .then(deleted7 => {
+        deleted7version = deleted7.version_;
+        expect(deleted7.deleted_).toBe(true);
```

Comment:
Hmm, is this actually a test? Are the tests run automatically by Travis? I was about to comment that this was a lot of code with no tests, but maybe this is it?

Reply:
Yes. All the `expect` calls in this file constitute the test expectations. These tests run on Travis (they were breaking before my follow-up commit!).